File: tic-tac-toe.py
def display_board(board):
    for row in board:
        print(" | ".join(row))
        print("-" * 5)

def is_winner(board, player):
    for row in board:
        if all(slot == player for slot in row):
            return True

    for col in range(3):
        if all(board[row][col] == player for row in range(3)):
            return True

    if all(board[i][i] == player for i in range(3)) or all(board[i][2 - i] == player for i in range(3)):
        return True

    return False

def is_draw(board):
    return all(all(slot != " " for slot in row) for row in board)

def play_game():
    board = [[" " for _ in range(3)] for _ in range(3)]
    current_player = "X"

    while True:
        display_board(board)
        row = int(input(f"Player {current_player}, choose your row (0-2): "))
        col = int(input(f"Player {current_player}, choose your column (0-2): "))

        if board[row][col] == " ":
            board[row][col] = current_player
            if is_winner(board, current_player):
                display_board(board)
                print(f"Player {current_player} wins!")
                break
            elif is_draw(board):
                display_board(board)
                print("The game is a draw!")
                break
            current_player = "O" if current_player == "X" else "X"
        else:
            print("That cell is already occupied, try again.")

if __name__ == "__main__":
    play_game()

File: tic-tac-toe.cpp
#include <iostream>
#include <vector>

using namespace std;

void renderBoard(const vector<vector<char>>& board) {
    for (const auto& row : board) {
        for (const auto& cell : row) {
            cout << cell << " ";
        }
        cout << endl;
    }
}

bool hasWinner(const vector<vector<char>>& board, char player) {
    for (const auto& row : board) {
        if (row[0] == player && row[1] == player && row[2] == player) {
            return true;
        }
    }

    for (int col = 0; col < 3; ++col) {
        if (board[0][col] == player && board[1][col] == player && board[2][col] == player) {
            return true;
        }
    }

    if (board[0][0] == player && board[1][1] == player && board[2][2] == player) {
        return true;
    }

    if (board[0][2] == player && board[1][1] == player && board[2][0] == player) {
        return true;
    }

    return false;
}

bool isTie(const vector<vector<char>>& board) {
    for (const auto& row : board) {
        for (const auto& cell : row) {
            if (cell == ' ') {
                return false;
            }
        }
    }
    return true;
}

int main() {
    vector<vector<char>> board(3, vector<char>(3, ' '));
    char currentPlayer = 'X';

    while (true) {
        renderBoard(board);
        int row, col;
        cout << "Player " << currentPlayer << ", enter the row (0, 1, or 2): ";
        cin >> row;
        cout << "Player " << currentPlayer << ", enter the column (0, 1, or 2): ";
        cin >> col;

        if (board[row][col] == ' ') {
            board[row][col] = currentPlayer;
            if (hasWinner(board, currentPlayer)) {
                renderBoard(board);
                cout << "Player " << currentPlayer << " wins!" << endl;
                break;
            } else if (isTie(board)) {
                renderBoard(board);
                cout << "The game is a tie!" << endl;
                break;
            }
            currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
        } else {
            cout << "That cell is already taken, please choose another." << endl;
        }
    }

    return 0;
}
